[Pipe] Add a `WithDevice` wrapper to specify device execution for a module.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65190

As described in https://github.com/pytorch/pytorch/issues/65093, there
could be modules which don't have any parameters/buffers. In this case, Pipe
determines that the module should be executed on CPU. However this might result
in unnecessary GPU to CPU transfers whereas the user expected the module to be
executed on the GPU itself by keeping its inputs and outputs on GPU.

For this use case, we introduce a `WithDevice` wrapper which can be used to
override which device a particular module should be executed on as part of the
pipeline.

#Closes: https://github.com/pytorch/pytorch/issues/65093

Differential Revision: [D31010027](https://our.internmc.facebook.com/intern/diff/D31010027/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @gcramer23

pritamdamania committed:
[Pipe] Add a `WithDevice` wrapper to specify device execution for a module.

As described in https://github.com/pytorch/pytorch/issues/65093, there
could be modules which don't have any parameters/buffers. In this case, Pipe
determines that the module should be executed on CPU. However this might result
in unnecessary GPU to CPU transfers whereas the user expected the module to be
executed on the GPU itself by keeping its inputs and outputs on GPU.

For this use case, we introduce a `WithDevice` wrapper which can be used to
override which device a particular module should be executed on as part of the
pipeline.

#Closes: https://github.com/pytorch/pytorch/issues/65093

Differential Revision: [D31010027](https://our.internmc.facebook.com/intern/diff/D31010027/)

[ghstack-poisoned]

mrshenli commented on test/distributed/pipeline/sync/test_pipe.py:
```diff
@@ -795,3 +794,26 @@ def forward(self, a, b):
+
+    model = nn.Sequential(fc1, WithDevice(dropout, 'cuda:1'))
+    model = Pipe(model, chunks=8)
+    model(torch.rand(16, 16).cuda(0))
```
do we expect the output to be on `cuda:1` as well? If so, we should test it?

mrshenli commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -133,6 +133,48 @@ def forward(self, *inputs):
 
+class WithDevice(nn.Module):
+    """
+    Wraps an nn.Module which is part of nn.Sequential passed into :class:`Pipe`
```
nit: format `nn.Module` and `nn.Sequential` as code.

mrshenli commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -133,6 +133,48 @@ def forward(self, *inputs):
+        >>> fc2 = nn.Linear(8, 4).cuda(1)
+        >>> dropout = nn.Dropout()
+        >>>
+        >>> # Dropout does not have any parameters/buffers, but we want to
```
I wonder what's the expected behavior here. If `dropout` does not have a device, should we just (recursively) inherit the device from the previous layer? So that users don't need to manually specify this?

SciPioneer replied:
+1

The goal here is to avoid the extra data transfer, so I also think no need to manually specify a device.

Another future use case where`WithDevice` may not work well is that, if the precursor layer is a tensor with lazy placement (`device="meta"`), then the current pure-computation layer can only be placed on a meta device, but I am afraid that the heuristic of co-locating two layers still won't be easily expressed by `WithDevice`.

pritamdamania87 replied:
I thought about this option as well, although this wouldn't work for the first partition in the sequence where we don't have any previous information to rely on. Also, I do feel it's probably better have an explicit API to deal with this rather than having many implicit rules.

For example, what if the user actually wanted a GPU to CPU transfer for a particular layer. Or if there was a sequence of layers with no params/buffers and they want them to run on different devices to spread out computation evenly?

pritamdamania87 replied:
> Another future use case whereWithDevice may not work well is that, if the precursor layer is a tensor with lazy placement (device="meta"), then the current pure-computation layer can only be placed on a meta device, but I am afraid that the heuristic of co-locating two layers still won't be easily expressed by WithDevice.

When the Pipe eventually initializes everything, we would have concrete devices for each partition and not "meta" at that point.

SciPioneer replied:
> e, what if the user actually wanted a GPU to CPU transfer for a particular layer. Or if there was a sequence of layers with no params/buffers and they want them to run on different devices to spread out computation evenly?



> I thought about this option as well, although this wouldn't work for the first partition in the sequence where we don't have any previous information to rely on. Also, I do feel it's probably better have an explicit API to deal with this rather than having many implicit rules.

Is it possible that the first layer is a pure computation layer? As long as the first layer in the first partition is not pure computation layer, shouldn't we always be able to find a valid device from the previous layers?

> For example, what if the user actually wanted a GPU to CPU transfer for a particular layer. Or if there was a sequence of layers with no params/buffers and they want them to run on different devices to spread out computation evenly?

Then that computation must be much more expensive than H2D ops. I agree `WithDevice` be more valuable in this case, but isn't this a uncommon use case? Maybe we can let the placement be the same as the previous layer by default, and provide `WithDevice` for extra flexibility?

pritamdamania87 replied:
> Is it possible that the first layer is a pure computation layer? As long as the first layer in the first partition is not pure computation layer, shouldn't we always be able to find a valid device from the previous layers?

It could be possible, for example we didn't anticipate that users would want a separate compute layer without any params/buffers as a different stage. Also, I think there is also a more general case where we could have a sequence of compute layers which need to be on different devices, in that case we do need `WithDevice`.

> Maybe we can let the placement be the same as the previous layer by default, and provide WithDevice for extra flexibility?

If we have to provide `WithDevice` anyways, I feel we should just have this instead of having a variety of implicit rules around how devices are handled.

pritamdamania committed:
Update on "[Pipe] Add a `WithDevice` wrapper to specify device execution for a module."


As described in https://github.com/pytorch/pytorch/issues/65093, there
could be modules which don't have any parameters/buffers. In this case, Pipe
determines that the module should be executed on CPU. However this might result
in unnecessary GPU to CPU transfers whereas the user expected the module to be
executed on the GPU itself by keeping its inputs and outputs on GPU.

For this use case, we introduce a `WithDevice` wrapper which can be used to
override which device a particular module should be executed on as part of the
pipeline.

#Closes: https://github.com/pytorch/pytorch/issues/65093

Differential Revision: [D31010027](https://our.internmc.facebook.com/intern/diff/D31010027/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/pipeline/sync/pipe.py:
```diff
@@ -149,7 +191,13 @@ def _split_module(modules: nn.Sequential) -> Tuple[List[nn.Sequential], List[tor
+            module = module.module
+            module.to(device)
+        else:
+            device = _retrieve_device(module)
```
nit: we can move the above into the `_retrieve_device` helper?
